File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/AddTwoNumbers.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.ListNode;

/**
 * @author zhenyu.jiang
 */
public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;
        int added = 0;
        while (l1 != null || l2 != null) {
            int val1 = l1 == null ? 0 : l1.val;
            int val2 = l2 == null ? 0 : l2.val;

            int sum = val1 + val2 + added;
            int actual = sum % 10;
            added = sum / 10;

            current.next = new ListNode(actual);
            current = current.next;
            l1 = l1 == null ? null : l1.next;
            l2 = l2 == null ? null : l2.next;
        }

        if (added == 1) {
            current.next = new ListNode(added);
        }

        return dummy.next;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/addTwoNumbers.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.ListNode

fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
    val dummy = ListNode(0)
    var current = dummy
    var carry = 0

    var node1 = l1
    var node2 = l2

    while (node1 != null || node2 != null || carry > 0) {
        // 计算当前位的和
        val sum = (node1?.`val` ?: 0) + (node2?.`val` ?: 0) + carry

        // 更新进位值
        carry = sum / 10

        // 创建新节点并移动指针
        current.next = ListNode(sum % 10)
        current = current.next!!

        // 移动输入链表指针
        node1 = node1?.next
        node2 = node2?.next
    }

    return dummy.next
}
